File: 2050.cpp
#include <stack>
#include <unordered_map>

struct Solution {
    int minimumTime(int n, const vector<vector<int>>& relations, const vector<int>& time) {
        unordered_multimap<int, int> edgeDict;
        vector<int> indegree(n, 0), preReqTime(n, 0);
        stack<int> st;
        for (const vector<int> &e : relations){
            edgeDict.insert(make_pair(e[0] - 1, e[1] - 1));
            ++indegree[e[1] - 1];
        }
        for (int i = 0; i < n; ++i){
            if (indegree[i] == 0){
                st.push(i);
            }
        }
        int res = -1, node, nei, timeCost;
        while (!st.empty()){
            node = st.top();
            st.pop();
            timeCost = preReqTime[node] + time[node];
            res = max(res, timeCost);
            auto rg = edgeDict.equal_range(node);
            for (auto it = rg.first; it != rg.second; ++it){
                nei = it->second;
                preReqTime[nei] = max(preReqTime[nei], timeCost);
                if (--indegree[nei] == 0){
                    st.push(nei);
                }
            }
        }
        return res;
    }
};

File: sol/2050/2050.py
from collections import defaultdict

class Solution:
    def minimumTime(self, n: int, relations: List[List[int]], time: List[int]) -> int:
        edgeDict = defaultdict(list)
        indegree = [0] * n
        for e in relations:
            edgeDict[e[0] - 1].append(e[1] - 1)
            indegree[e[1] - 1] += 1
        stack = list(i for i in range(n) if indegree[i] == 0)
        res = -1
        preReqTime = [0] * n
        while len(stack) != 0:
            node = stack.pop()
            timeCost = preReqTime[node] + time[node]
            res = max(res, timeCost)
            for nei in edgeDict[node]:
                preReqTime[nei] = max(preReqTime[nei], timeCost)
                indegree[nei] -= 1
                if indegree[nei] == 0:
                    stack.append(nei)
        return res
